File: src/2024/day11.js
const input = [872027, 227, 18, 9760, 0, 4, 67716, 9245696];
//const input = [125, 17];
//const input = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
//const input = [0]
let count = input.length;

// this seem to stop working after 22 digits, hopefully it will be ok :/
function nbDigits(num) {
  return String(num).length;
}

const map = new Map();
// i = iteration
// n = number
// return number of numbers created after "i" iterations
function p(n, i) {

  if (i == 0) return 1;

  const key = `${n}-${i}`;
  const v = map.get(key);
  if (v != undefined) {
    return v;
  }


  let res = 0;
  let nbD = nbDigits(n);
  if (n === 0) {
    res = p(1, i - 1);
  } else if ((nbD & 1) == 0) {
    const shift = Math.pow(10, nbD / 2)
    res = p(Math.trunc(n / shift), i - 1) + p(n % shift, i - 1);
  } else {
    res = p(n * 2024, i - 1);
  }
  map.set(key, res);
  return res;
}

let res = 0;
for (let bal of input) {
  res += p(bal, 75);
}
console.log(res);

for (let i = 0; i < 75; i++) {
  let newCount = 0;
  for (let j = 0; j < count; j++) {
    const n = input.shift();
    if (n == 0) {
      newCount += 1;
      input.push(1);
      continue;
    }
    let nbD = nbDigits(n);
    if ((nbD & 1) == 0) {
      const shift = Math.pow(10, nbD / 2)
      input.push(Math.trunc(n / shift));
      input.push(n % shift);
      newCount += 2;
      continue;
    }
    input.push(n * 2024);
    newCount += 1;
  }
  count = newCount;
}

console.log(input.length);

File: src/2024/day11.zig
const std = @import("std");
const mem = std.mem;

input: []const u8,
allocator: mem.Allocator,

fn nbDigits(n: u64) u64 {
    if (n == 0) return 0;
    return std.math.log10(n) + 1;
}

fn p(map: *std.StringHashMap(u64), n: u64, i: u64, allocator: mem.Allocator) u64 {
    if (i == 0) return 1;

    const key = std.fmt.allocPrint(allocator, "{d}-{d}", .{ n, i }) catch unreachable;
    if (map.contains(key)) {
        return map.get(key).?;
    }
    var res: u64 = 0;
    const nbD = nbDigits(n);
    if (n == 0) {
        res = p(map, 1, i - 1, allocator);
    } else if ((nbD & 1) == 0) {
        const shift = std.math.pow(u64, 10, nbD / 2);
        res = p(map, n / shift, i - 1, allocator) + p(map, n % shift, i - 1, allocator);
    } else {
        res = p(map, n * 2024, i - 1, allocator);
    }
    map.put(key, res) catch unreachable;
    return res;
}

pub fn part1(this: *const @This()) !?i64 {
    // last character is annoying
    var it = mem.tokenizeScalar(u8, this.input[0 .. this.input.len - 1], ' ');
    var map = std.StringHashMap(u64).init(this.allocator);
    defer {
        var mit = map.iterator();
        while (mit.next()) |entry| {
            this.allocator.free(entry.key_ptr.*);
        }
        map.deinit();
    }

    var res: u64 = 0;
    const i: u64 = 25;
    while (it.next()) |n| {
        const num = std.fmt.parseInt(u64, n, 10) catch {
            continue;
        };
        res += p(&map, num, i, this.allocator);
    }
    return @intCast(res);
}

pub fn part2(this: *const @This()) !?i64 {
    // last character is annoying
    var it = mem.tokenizeScalar(u8, this.input[0 .. this.input.len - 1], ' ');
    var map = std.StringHashMap(u64).init(this.allocator);
    defer {
        var mit = map.iterator();
        while (mit.next()) |entry| {
            this.allocator.free(entry.key_ptr.*);
        }
        map.deinit();
    }

    var res: u64 = 0;
    const i: u64 = 75;
    while (it.next()) |n| {
        if (n[0] < '0' or n[0] > '9') continue;
        const num = std.fmt.parseInt(u64, n, 10) catch unreachable;
        res += p(&map, num, i, this.allocator);
    }
    return @intCast(res);
}

test "it should do nothing" {
    const allocator = std.testing.allocator;
    const input = "";

    const problem: @This() = .{
        .input = input,
        .allocator = allocator,
    };

    try std.testing.expectEqual(null, try problem.part1());
    try std.testing.expectEqual(null, try problem.part2());
}
